src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -906,10 +906,28 @@ private bool IsPossibleAttributeDeclaration()
 
         private SyntaxList<AttributeListSyntax> ParseAttributeDeclarations(bool inExpressionContext)
         {
-            var attributes = _pool.Allocate<AttributeListSyntax>();
             var saveTerm = _termState;
             _termState |= TerminatorState.IsAttributeDeclarationTerminator;
 
+            // An attribute can never appear *inside* an attribute declaration in the language.  However, during parsing
+            // we can end up in a state where we're trying to exactly that, through a path of
+            // Attribute->Argument->Expression->Attribute (since attributes can not be on lambda expressions).
+            //
+            // Worse, when we are in a deeply ambiguous (or gibberish) scenario, where we see lots of code with
+            // `... [ ... [ ... ] ... ] ...` in it, we can get into exponential speculative parsing blowup.  Where we
+            // try `[ ... ]` both as an attribute *and* a collection expression (including needing to speculate on the
+            // nested `[ ... ]`s we run into.
+            //
+            // This speculative blowup can be very bad, effectively live-locking a host while each successive, nested,
+            // speculative parse gets slower and slower.  Since we cannot ever legally have an attribute within an
+            // attribute, we bail out here immediately syntactically.  This does mean we won't parse something like:
+            // `[X([Y]() => {})]` without errors.  But as that's not semantically legal anyways, it's better to not
```

Comment:
Do we have any other examples where the parser intentionally reports errors for syntax that is valid according to the grammar?

Reply:
yes.  `var a = b/*c` is legal as per the grammar.  It should be parsed as "b divided by the deference of c".  i.e. `var a = b/(*c)`.  However, supporting that correct grammatical interpretation is to onerous.  

Specifically, if we did that, we'd have to then keep track that this *could* be the start of a comment, and then figure out if any edit later on (like inserting a `/` or `*`, or making any textual changes that could cause a `*/` to come into existence), so that we could then jump back all the way there and now reinterpet the `/*` as the start of a comment instead.

We decided this was just too computationally expensive, and so difficult to incorporate into a incremental parsing model that we just explicitly choose to get it wrong *unilaterally*.  In other words we *always* see `/*` as the start of a comment, even if it isn't a comment, and should actually be interpreted as a binary and unary expression.

Reply:
The compiler has broad leeway in how to handle illegal or near-impossible-to-handle cases.  While producing a syntax tree without diagnostics is nice-to-have, it's not a requirement.  And in error conditions, we can decide what diagnostics to produce and from where they'll come from.

Note: i'm not opposed to an approach that produces a valid tree here, while not having exponential blowup.  I just don't know what such an approach would be.  If someone can figure that out, then i'm all for it.  I'm just trying to get us into a situation where legal code continues to compile, illegal code continues to give error diagnostics, and we don't have the blowup we currently have.

This PR does degrade the error messages slightly in a hugely uncommon use case, while preventing literal deadlocks.  Seems like a reasonable tradeoff for me :)